ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/VirtualArtifactRepository.java

```diff
@@ -0,0 +1,215 @@
+
+import aQute.bnd.osgi.Jar;
+import aQute.bnd.osgi.WriteResource;
+
+/**
+ * In-Memory representation of a artifact repository based on a non-IU target
+ * location. This repository is used during the planner resolution of an IU
+ * target location to supply artifacts from other (non-IU) locations.
+ */
+@SuppressWarnings("restriction")
+public class VirtualArtifactRepository extends AbstractArtifactRepository {
+	private static final String NAME = "Non-IU Artifact Repository @ "; //$NON-NLS-1$
+	private static final String DESCRIPTION = """
+			In-Memory repository created for a single Non-IU repository, used
+			during the planner resolution of a real IU repository.
+			"""; // $NON-NLS-1$ //$NON-NLS-1$
+	private static final String MEMORY_PREFIX = "memory://"; //$NON-NLS-1$
+	// BundleInfo or IFeatureModel
+	private final Map<IArtifactDescriptor, Object> artifacts = new HashMap<>();
+
+	public VirtualArtifactRepository(IProvisioningAgent agent, ITargetLocation targetLocation) {
+		super(agent, NAME + getLocationSafe(targetLocation), targetLocation.getType(), null,
+				URI.create(MEMORY_PREFIX + UUID.randomUUID()), DESCRIPTION + '\n' + targetLocation.serialize(), null,
+				null);
+		Assert.isTrue(targetLocation.isResolved());
+		for (TargetBundle targetBundle : targetLocation.getBundles()) {
+			if (!targetBundle.getStatus().isOK()) {
+				PDECore.log(Status.warning(NLS.bind(Messages.VirtualArtifactRepository_0, targetBundle)));
+				continue;
+			}
+			BundleInfo bundleInfo = targetBundle.getBundleInfo();
+			IArtifactKey artifactKey = BundlesAction.createBundleArtifactKey(bundleInfo.getSymbolicName(),
+					bundleInfo.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, bundleInfo);
+		}
+		for (TargetFeature targetFeature : targetLocation.getFeatures()) {
+			IArtifactKey artifactKey = FeaturesAction.createFeatureArtifactKey(targetFeature.getId(),
+					targetFeature.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, targetFeature.getFeatureModel());
+		}
+	}
+
+	private static String getLocationSafe(ITargetLocation targetLocation) {
+		try {
+			return targetLocation.getLocation(false);
+		} catch (CoreException e) {
+			return "<unknown>"; //$NON-NLS-1$
+		}
+	}
+
+	@Override
+	public IStatus getRawArtifact(IArtifactDescriptor descriptor, OutputStream destination, IProgressMonitor monitor) {
+		Object artifactModel = artifacts.get(descriptor);
+		if (artifactModel == null) {
+			return Status.error(NLS.bind(Messages.VirtualArtifactRepository_1, descriptor));
+		}
+		try (InputStream is = getArtifactInputStream(artifactModel)) {
+			is.transferTo(destination);
+			return Status.OK_STATUS;
+		} catch (Exception e) {
+			return Status.error(e.getLocalizedMessage(), e);
+		}
+	}
+
+	private InputStream getArtifactInputStream(Object artifactModel) throws Exception {
```

Comment:
This method would probably be simple, if in the caller to destination `OutputStream` would just be passed as an argument and if in this method the artifact would be written directly to the output stream.
For example in the last case, this would avoid creating an `InputStream` where the content is first written to an ByteArrayOutputStream.
Or in case there is a `featureJar` `File` you could just use `Files.copy(featureJar.toPath, destition)` to copy the file content to the output stream.